Support GitHub Enterprise

castler opened the pull request:
GitHub can be hosted by companies for their own usage (under their own URL).
This changset removes the hard-coded url to `github.com` and makes it confligurable for a user, which GitHub-instances shall be supported.

Closes #134

castler committed:
Support GitHub Enterprise

GitHub can be hosted by companies for their own usage (under their own
URL).
This changset removes the hard-coded url to `github.com` and makes it
confligurable for a user, which GitHub-instances shall be supported.

Closes #134

castler commented:
@mikelalcon @hsudhof

Sorry to bother.
Is there any chance of getting this, or similar versions, merged? Otherwise I would close the PR - to not have unfinished work.

Thanks for your feedback!

chris-campos commented on java/com/google/copybara/git/GitModule.java:
```diff
@@ -341,6 +342,16 @@ public GitOrigin origin(
+        GitHubOptions githubOptions = options.get(GitHubOptions.class);
+        GitHubHost ghHost = githubOptions.getGitHubHost(url);
+        isGitHubUrl = ghHost.isGitHubUrl(url);
+    } catch (EvalException e) {
```
empty catch blocks are a bit of an anti-pattern, and I'd like to avoid having these if possible. is there any way you can refactor this into a method in GitHubOptions that returns whether a url belongs to one of the GH hosts?

if you must keep the empty catch block, please leave a comment explaining why we catch and swallow the exception

chris-campos commented on java/com/google/copybara/git/GitModule.java:
```diff
@@ -3045,28 +3061,34 @@ private String fixHttp(String url, Location location) {
+                  !ghHost.isGitHubUrl(url),
+                  "Git origins with github should use github approval providers!");
+      } catch(EvalException e){
+         // nothing to-do, apparently this is no GitHub URL.
```
same comments as above

chris-campos commented on javatests/com/google/copybara/git/GitOriginTest.java:
can we add a test that checks if a origin URL is considered a valid GitHub URL when that host is set as a GitHub host?

chris-campos reviewed (changes requested):
TY for the contribution! Please see line comments.

castler commented:
Great, I will try to solve that in the next days!

castler commented:
@chris-campos tried to apply your comments in:

https://github.com/google/copybara/compare/c3f828295d778da8104501c0bd2b0d8e9305eb11..3f870864744eb257048bbe6708ad95806a9f6357

chris-campos commented on java/com/google/copybara/git/GitHubOptions.java:
```diff
@@ -86,6 +89,27 @@ public class GitHubOptions implements Option {
+      names = "--github-allowed-hosts",
+      description = "If using GitHub Enterprise, one needs to specify valid hosts. By default only `github.com` is supported."
+  )
+  public List<String> gitHubAllowedHosts = ImmutableList.of("github.com");
```
Thinking about this more, I'm realizing that a flag to set the "allowed" GitHub hosts is not going to scale well for our internal use, and it might not be user friendly for our OSS users. We should avoid using a flag for setting the host allowlist.

 I see why the flag was needed in the first place: `git.origin` was allowed to accept GitHub use cases, which in retrospect was not ideal. However, we also have `git.github_origin` (GitHub specific) in the module, which gives me an idea.

What if we make it so that `git.github_origin` assumes that the host in the URL _is a valid GitHub host_? i.e., if a user says the repo URL is `foo.com`, then we just treat it as if it has GitHub APIs? 

As for the standard `git.origin`, `github.com` can be considered the only valid GitHub host. Therefore, Github enterprise specific origins should use `git.github_origin`

In this way, we are:
- leveraging the user's intent in choosing `github_origin` to set it as a valid GitHub host as that is the most likely intent
- preserving the existing behavior for legacy `git.origin` users.

what do you think? please LMK if you have any questions, and thank you again for working on this.

castler replied:
Understood - I am afraid of cases where no `github` specific function is available. I am not that familiar yet with CopyBara in order to judge this out of my head if such cases exist. 

So I will give this a try in the next days.

chris-campos replied:
@castler Hello! 

I just wanted to follow up on this and how your progress is going on this change. We are really interested in adding this feature into Copybara. 

If you can no longer work on it, do you mind if we clone your change and continue work on it? We are also happy to answer any questions you may have. 

Thanks again for this great contribution thus far!

chris-campos replied:
Thank you, I will see if I can make the changes I suggested above.

chris-campos reviewed (commented):
Thanks for the code changes! Left one more comment about the added flag, LMK what you think

chris-campos committed:
Allow GitHub origin to accept any host

To better support GitHub enterprise, the GitHub Origin will just assume any host it's given is a GitHub or Github enterprise host.

GitOrigin will continue to accept GitHub.com URLs.

chris-campos commented:
@sainathseelam Hi there! We are now working on this internally.

castler commented:
Thanks for integrating the changes - it seems to work for us :)
